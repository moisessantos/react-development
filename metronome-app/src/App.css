.metronome {
  text-align: center;
  max-width: 375px;
  margin: 0 auto;
  padding: 30px;
}

.slider {
  width: 100%;
  margin: 10px;
}

.h3 {
  margin-bottom: 1rem !important;
}