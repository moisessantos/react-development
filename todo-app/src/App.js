import React from 'react';
import TodoList from './components/TodoList';

const App = () => <TodoList items={['Turn off the alarm', 'Have breakfast']} />;

export default App;
