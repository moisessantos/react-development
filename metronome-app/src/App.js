import React, { Component } from 'react';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import Slider from '@material-ui/lab/Slider';

import click1 from './sounds/click1.wav';
import click2 from './sounds/click2.wav';

import './App.css';

class Metronome extends Component {
  constructor(props) {
    super(props);

    this.state = {
      playing: false,
      count: 0,
      bpm: 100,
      beatsPerMeasure: 4
    };

    this.click1 = new Audio(click1);
    this.click2 = new Audio(click2);
  }
  
  startStop = () => {
    if (this.state.playing) {
      // Stop the timer
      clearInterval(this.timer);
      this.setState({
        playing: false
      });
    } else {
      // Start a timer with the current BPM
      this.timer = setInterval(
        this.playClick,
        (60 / this.state.bpm) * 1000
      );
      this.setState(
        {
          count: 0,
          playing: true
          // Play a click "immediately" (after setState finishes)
        },
        this.playClick
      );
    }
  };

  playClick = () => {
    const { count, beatsPerMeasure } = this.state;
  
    // The first beat will have a different sound than the others
    if (count % beatsPerMeasure === 0) {
      this.click2.play();
    } else {
      this.click1.play();
    }
  
    // Keep track of which beat we're on
    this.setState(state => ({
      count: (state.count + 1) % state.beatsPerMeasure
    }));
  };

  handleBpmChange = (event, value) => {
    const bpm = value;
  
    if (this.state.playing) {
      // Stop the old timer and start a new one
      clearInterval(this.timer);
      this.timer = setInterval(this.playClick, (60 / bpm) * 1000);
  
      // Set the new BPM, and reset the beat counter
      this.setState({
        count: 0,
        bpm
      });
    } else {
      // Otherwise just update the BPM
      this.setState({ bpm });
    }
  };

  render() {
    const { playing, bpm } = this.state;

    return (
      <div className="metronome">
        <div className="bpm-slider">
          <Typography variant="h3" className="h3">{bpm} BPM</Typography>
          <Slider 
            type="range" 
            className="slider"
            min={60} 
            max={300} 
            value={bpm} 
            step={5}
            onChange={this.handleBpmChange} />
        </div>
        
        <Button variant="contained" color="primary" onClick={this.startStop}>
          {playing ? 'Stop' : 'Start'}
        </Button>
      </div>
    );
  }
}

export default Metronome;