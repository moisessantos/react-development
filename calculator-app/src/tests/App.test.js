import React from 'react';
import { shallow } from 'enzyme';
import App from '../App';

describe('App tests', () => {
    it('should render as expected', () => {
        const wrapper = shallow(<App />);
        expect(wrapper).toMatchSnapshot();
    });

    it.each(
        [
            ["calculate", "="],
            ["reset","C"],
            ["backspace", "CE"]
        ],
        'onClick should call %s when %s is pressed'
        ,(func, val) => {
            const wrapper = shallow(<App />);
            const spy = jest.spy(wrapper.instance(), func);

            wrapper.instance().onClick(val);

            expect(spy).toBeCalledTimes(1);
        });
    
    it('onClick should append to state when a number or operator is pressed', () => {
        const wrapper = shallow(<App />);
        wrapper.setState({ result: "1" })
        wrapper.instance().onClick("+");

        expect(wrapper.state().result).toBe("1+");
    });

    it('reset should reset the result', () => {
        const wrapper = shallow(<App />);
        wrapper.setState({ result: "1" })
        wrapper.instance().reset();

        expect(wrapper.state().result).toBe("");
    });

    
    it('backspace should remove the last char from the result', () => {
        const wrapper = shallow(<App />);
        wrapper.setState({ result: "1+" })
        wrapper.instance().backspace();

        expect(wrapper.state().result).toBe("1");
    });

    it('calculate should do the calculation in the result', () => {
        const wrapper = shallow(<App />);
        wrapper.setState({ result: "1+2" })
        wrapper.instance().calculate();

        expect(wrapper.state().result).toBe("3");
    });

    it('calculate should do the calculation when -- is present in the result', () => {
        const wrapper = shallow(<App />);
        wrapper.setState({ result: "1--2" })
        wrapper.instance().calculate();

        expect(wrapper.state().result).toBe("3");
    });

    it('calculate should return error in case of typos', () => {
        const wrapper = shallow(<App />);
        wrapper.setState({ result: "1+*2" })
        wrapper.instance().calculate();

        expect(wrapper.state().result).toBe("error");
    });
})