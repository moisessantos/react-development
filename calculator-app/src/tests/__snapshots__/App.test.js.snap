// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests should render as expected 1`] = `
<div>
  <div
    className="calculator-body"
  >
    <h1>
      Simple Calculator
    </h1>
    <Result />
    <Memo(KeyPad)
      handleClick={[Function]}
    />
  </div>
</div>
`;
